Resolve #533 & #534 -- EditBuff added, buff slot value mgmt implemented

HarmfulBreeze opened the pull request:
Everything has been tested and is working fine.
Open to all suggestions.

HarmfulBreeze committed:
Implemented EditBuff packet and added buff slot management

HarmfulBreeze committed:
Changed the loop variable type to byte

Thanks for the suggestion

MythicManiac commented on GameServerLib/Logic/GameObjects/Unit.cs:
```diff
@@ -739,6 +744,36 @@ public ClassifyUnit ClassifyTarget(Unit target)
+                if (!AppliedBuffs.Values.Contains(buff))
+                    continue;
+
+                for (byte i = 1; i <= AppliedBuffs.Count + 1; i++)
```
```C#
foreach (var kvp in AppliedBuffs)
{
    if (kvp.Value.Equals(buff)) {
        AppliedBuffs.Remove(kvp.Key);
    }
}
```

HarmfulBreeze replied:
Can't use a foreach loop since it would modify the collection and crash on the next iteration.

MythicManiac commented on GameServerLib/Logic/GameObjects/Unit.cs:
```diff
@@ -739,6 +744,36 @@ public ClassifyUnit ClassifyTarget(Unit target)
+            {
+                if (AppliedBuffs.ContainsKey(i))
+                    continue;
+                AppliedBuffs.Add(i, b);
```
Why are we updating buff slots in a function called `GetBuffSlot`?

At the very least make the function name imply what it does, `UpdateBuffSlots` for example. Also would it not make more sense to remove the buffs that you want to remove in the `RemoveBuffSlot` function instead of here? Do we really need a `BuffSlotToRemove` variable?

MythicManiac reviewed (commented):
Got some things you could address, but nice work overall 👍

HarmfulBreeze committed:
Moved the buff slot removal to RemoveBuffSlot

MythicManiac reviewed (commented):
Instead of having copy-paste code in two separate locations (Line 747 and 763) write a function that does what you want, something like the following:

```C#
public void GetBuffSlot(Buff b)
{
    Buff found;
    foreach(var kvp in AppliedBuffs)
    {
        if(kvp.Value.Equals(b)) {
            found = kvp.Key;
        }
    }
    return found;
}
```

Iterating over dictionaries should always be done by the dictionary's key set (which `foreach` here does), as you have no guarantee of what the keys can be. For example, imagine if you have the following dictionary:
```C#
var dict = new Dictionary<int, string>() {
    {-5, "hello"},
    {4952, "world"}
}
```
the current method of looping over it would never even see the `-5` index, and would take a huge amount of cycles to reach `4952`.

HarmfulBreeze committed:
Merge branch 'indev' into indev

HarmfulBreeze committed:
Renamed GetBuffSlot() to GetNewBuffSlot() and added GetBuffSlot()

HarmfulBreeze committed:
Moved duplicate code to GetBuffSlot()

HarmfulBreeze committed:
Renamed BuffSlotFlag to BuffSlotType (thanks for the explanation)

MythicManiac reviewed (commented):
Could `AppliedBuffs` be an array or a list instead of a dictionary, seeing as you intend to use it like one either way?

Doesn't need to be changed, just want to hear your thoughts.

HarmfulBreeze commented:
I think that here using a Dictionary seems better:
- For the K/V system that corresponds perfectly to our situation where you match a slot and a buff
- From what I've read, it's faster to look stuff up by using a Dictionary in our situation

So yeah, using a Dictionary here seems to be the best option.

MythicManiac commented:
The thing is you're using the dictionary like you would use a list/array, so you're actually negatively impacting the performance.

Array is a fixed size pre-initialized set of items (like we have here from what I've understood), while Dictionary is a dynamic mapping. If you need to be able to find empty slots by iterating keys as if the Dictionary was an array, you probably want to use an array

HarmfulBreeze committed:
Changed AppliedBuffs' type to Array
We also now call RemoveBuffSlot after having notified the client to avoid possible overlapping issues

MythicManiac commented on GameServerLib/Logic/GameObjects/Unit.cs:
```diff
@@ -751,6 +758,46 @@ public ClassifyUnit ClassifyTarget(Unit target)
+            }
+        }
+
+        private byte GetBuffSlot(Buff b, BuffSlotType type)
```
This function could still be simplified a bit

What I originally wished for when you merged your two separate functions into one was for them to share logic. Right now it's essentially two different functions inside the same body.

Compare to the following example:
```C#
private byte GetBuffSlot(BuffSlotType type, Buff buff=null)
{
    for (byte i = 0; i < AppliedBuffs.Length; i++) // Find the first open slot
    {
        if (type == BuffSlotType.New && AppliedBuffs[i] == null)
        {
            return i; // If the slot is open and an open one was requested, return this slot
        }
        if (type == BuffSlotType.Existing && buff.Equals(b))
        {
            return i; // Returns the slot corresponding to the 
        }
    }
    throw new Exception("No slot found");
}
```
Less code used and cleaner, don't you agree?

This could be simplified even more if you scratch the `BuffSlotType` entirely, and instead use the buff parameter to decide what we're looking for, like so:

```C#
private byte GetBuffSlot(Buff buffToLookFor=null)
{
    for (byte i = 0; i < AppliedBuffs.Length; i++) // Find the first open slot
    {
        // I'm not sure if this matches buff.Equals(buffToLookFor) like you had originally,
        // so be sure to test for that
        if (AppliedBuffs[i] == buffToLookFor)
        {
            return i;
        }
    }
    throw new Exception("No slot found with requested value");
}
```

As always, let me know your thoughts 😄

HarmfulBreeze replied:
Hey, I like the first solution. I now understand what you meant back then. 🤔
But the second solution really is awesome. I love it! I'll check if ``==`` works but I'm pretty sure it will.
I find this default parameter value usage really smart, it makes the code super small and clean. 

Shall we still return ``0x00`` if we're looking for a Buff's slot, or shall we simply throw?
Thanks!

MythicManiac replied:
I think I'd have the return type an `int` and return `-1` if nothing was found, thus allowing the error handling to be done elsewhere. No "right" choice here however, up to preference.

HarmfulBreeze replied:
Yeah seems good to me. I'll test it asap. By the way, IIRC, you can't have 0x00 as a buff slot, so I'll have to change that too. That's why I originally returned 0x00 when buff isn't applied

MythicManiac replied:
If that's the case, 0x00 is alright too. Just as long as it's always the same and outside of the range of valid slots

MythicManiac reviewed (commented):
The Dictionary->Array change you did looks good, nice work 👍

There is still one thing I think could use improving 😄 I'm fine with merging it as is, and create an issue about improving this part of the code, as I'm the one who suggested you'd format your code this way without explaining clearly enough the intention of it.

Anyway, up to you whether you want to change it or not, good job on this PR so far 👍

HarmfulBreeze committed:
Final change to GetBuffSlot
Using for-loop starting at 1 to avoid writing to slot 0x00

HarmfulBreeze committed:
Fixed submodules not being able to be fetched (thanks Dniel8888)

MythicManiac reviewed (approved):
Looking a lot better now, nice work 👍
